### PR TITLE
Housekeeping for Canvas.py

### PR DESCRIPTION
**Pull request by Nusnus**
Small housekeeping for `canvas.py` for the 5.3 release.
Will keep it in Draft until ready for merge.
Comments are welcome for any commit in the PR.

### CODE REVIEW

**Comment by Nusnus** [comment]
PR is ready for merge from my side.

File: celery/canvas.py [hunk 1]
```diff
@@ -2248,13 +2247,14 @@ def link_error(self, errback):
...
         """
         if self.app.conf.task_allow_error_cb_on_chord_header:
-            # self.tasks can be a list of the chord header workflow.
-            if isinstance(self.tasks, (list, tuple)):
-                for task in self.tasks:
-                    task.link_error(errback)
-            else:
-                self.tasks.link_error(errback)
+            for task in self.tasks:
+                task.link_error(errback)
```
Comment (thedrow) [inline celery/canvas.py:2251]: Why did you remove this condition? `self.tasks` may be a list or a tuple and if it isn't, it's a group.

E.g.
```
chord([s1.s(), s2.s()], body.s())
chord(group(s1.s(), s2.s()), body.s())
```

We need to support both cases.

File: t/unit/tasks/test_canvas.py [hunk 2]
```diff
@@ -454,6 +455,9 @@ def test_OR(self):
         assert isinstance(ax, _chain)
         assert len(ax.tasks), 3 == 'consolidates chain to chain'
 
+        with pytest.raises(TypeError):
+            assert signature('foo') | None
+
```
Comment (thedrow) [inline t/unit/tasks/test_canvas.py:460]: Please add a subtest that explains this assert.
There's no need to add subtests for all the cases but if we're adding a new assert, we should describe why it is there.

**Comment by Nusnus** [comment]
@auvipy If you have a patch to fix the linter this could be a good time to add it in, so we can close open ends for the upcoming release.

File: t/unit/tasks/test_canvas.py [hunk 3]
```diff
@@ -2386,3 +2395,63 @@ def test_is_dict(self):
...
+        d2 = {'dict': {'b': 2}, 'list': [3, 4], 'set': {'a', 'b'}}
+        _merge_dictionaries(d1, d2)
+        assert d1 == {
+            'dict': {'a': 1, 'b': 2},
+            'list': [1, 2, 3, 4],
+            'tuple': (1, 2),
+            'set': {'a', 'b'}
+        }
+
+    @pytest.mark.parametrize('d1,d2,expected_result', [
```
[the remark this answers is not in the document]
Comment (thedrow) [reply t/unit/tasks/test_canvas.py:2451]: Looks good

**Review by thedrow — CHANGES_REQUESTED** [review]

File: celery/canvas.py (hunk 1 above)
Comment (Nusnus) [reply celery/canvas.py:2251]: There's no problem with that, `self.tasks` is accessed correctly with both examples.
To prove it, I've added the PR a test case with your both examples: `test_link_error_on_chord_header()`
Check it out to confirm and resolve conversation if you're satisfied @thedrow

File: t/unit/tasks/test_canvas.py (hunk 2 above)
Comment (Nusnus) [reply t/unit/tasks/test_canvas.py]: Done

**Comment by Nusnus** [comment]
All comments have been addressed, PR is ready for merge from my side.

**Comment by Nusnus** [comment]
> PR is ready for merge from my side

This only means I have responded to all of the open ends from the PR, to mark "I have nothing else to do here".
So no pressure, but if all of the comments are resolved and fixed, then we shouldn't wait for the sake of waiting :)

**Comment by Nusnus** [comment]
I don’t understand what's wrong with the integration tests, they just broke and the tests are not even running, and I'm not sure how to fix it.
Do you have any advice?
@auvipy @thedrow

**Comment by auvipy** [comment]
can you please check which version of tox the CI is using now? tox got a new major release very recently AFAIK.

**Comment by Nusnus** [comment]
The integration tests were running an old version, you were right @auvipy .
I've added a fix, let's hope everything works now 🙏

**Comment by auvipy** [comment]
will do a thorough review tomorrow

**Review by thedrow — APPROVED** [review]
LGTM
